[clang][analyzer] Support `ownership_{returns,takes}` attributes

pskrgag opened the pull request:
Add support for checking mismatched ownership_returns/ownership_takes attributes.

Closes #76861

pskrgag committed:
csa/tests: add test suite for custom allocation classes

pskrgag commented:
Hello @steakhal,

could you, please, take a look? I just found linked issue quite interesting and decided to give it a try. I am new to llvm world, so I would really appreciate your review

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -103,14 +103,46 @@ using namespace std::placeholders;
+
+class AllocationFamily {
+public:
+  AllocationFamily(AllocationFamilyKind kind,
```
I'd suggest marking this `explicit`.

pskrgag replied:
Marking with `explicit` would cause compile errors for code like `family == AF_Smth`. We could wrap all `AF_*` with direct constructor call, but, I think, it would be less clean. What do you think?

steakhal replied:
I'd prefer having this explicit.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -103,14 +103,46 @@ using namespace std::placeholders;
+  }
+
+  bool operator==(const AllocationFamily &Other) const {
+    if (Other.kind() == this->kind()) {
+      return this->kind() == AF_Custom ? this->_customName == Other._customName
+                                       : true;
+    } else {
+      return false;
+    }
```
I don't see why a `return std::tie(_kind, _customName) == std::tie(Other._kind, Other._customName)` wouldn't work.
I find that more conventional to the one you proposed.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -103,14 +103,46 @@ using namespace std::placeholders;
+    }
+  }
+
+  bool operator==(const AllocationFamilyKind &kind) {
+    return this->kind() == kind;
+  }
```
Can we drop this?

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -103,14 +103,46 @@ using namespace std::placeholders;
+    return this->kind() == kind;
+  }
+
+  bool operator!=(const AllocationFamilyKind &kind) { return !(*this == kind); }
```
Could you define this for the class type instead?

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -103,14 +103,46 @@ using namespace std::placeholders;
+
+  bool operator!=(const AllocationFamilyKind &kind) { return !(*this == kind); }
+
+  std::optional<StringRef> name() const { return _customName; }
+  AllocationFamilyKind kind() const { return _kind; }
+
+private:
+  AllocationFamilyKind _kind;
+  std::optional<StringRef> _customName;
```
I think it should be fine to have these fields as public members, given the complexity of this class. This is basically a std::pair. Feel free to use `struct` instead of `class`.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -194,7 +226,7 @@ class RefState {
     ID.AddInteger(K);
     ID.AddPointer(S);
-    ID.AddInteger(Family);
+    ID.AddInteger(Family.kind());
```
This Profile wouldn't take the custom allocation function name into account.
I'd suggest letting the Family define its own Profile method and just pass the ID to it.
That is the canonical way of doing this.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1670,15 +1702,18 @@ MallocChecker::MallocMemReturnsAttr(CheckerContext &C, const CallEvent &Call,
 
-  if (Att->getModule()->getName() != "malloc")
-    return nullptr;
+  // Preseve previous behavior when "malloc" class means AF_Malloc
+  auto attrClassName = Att->getModule()->getName();
+  auto AF = attrClassName == "malloc"
+                ? AF_Malloc
+                : AllocationFamily(AF_Custom, attrClassName);
```
You could do this in the ctor of `AllocationFamily`. That would ensure that it's used like that consistently across all sites.

I've seen a variable name `Family` for representing the same value before. Have you considered using the same variable name to boost consistency and readability across the patch?

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1877,6 +1914,33 @@ static bool didPreviousFreeFail(ProgramStateRef State,
+    if (!FD)
+      return;
+
+    if (FD->hasAttrs()) {
```
You could switch this into an early-return to reduce indentation for the rest of the code.
Same applies to the `if (const CallExpr *CE = dyn_cast<CallExpr>(E))` a few lines above.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1877,6 +1914,33 @@ static bool didPreviousFreeFail(ProgramStateRef State,
+          else
+            os << ", which takes ownership of ";
+
+          os << '\'' << I->getModule()->getName() << "'";
```
You use single quoted char and double quoted char. I just highlight this inconsistency.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1918,26 +1982,54 @@ static bool printMemFnName(raw_ostream &os, CheckerContext &C, const Expr *E) {
+    return;
+  case AF_Alloca:
+  case AF_None:
+    llvm_unreachable("not a deallocation expression");
```
I'd suggest swapping this `llvm_unreachable` to an `assert(false && "not a deallocation expression"); return;` as this code is probably not in the hot path, thus the use of `llvm_unreachable` is unjustified.

pskrgag replied:
I think, all `llvm_unreachable` s  in this file could be turned in `assert` , since they just check sanity of calling contract. Will clean up all of them in separate patch, thanks!

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1918,26 +1982,54 @@ static bool printMemFnName(raw_ostream &os, CheckerContext &C, const Expr *E) {
+    return;
+  case AF_Alloca:
+  case AF_None:
+    llvm_unreachable("suspicious argument");
```
Same goes here, use a regular `assert` and `return`.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3483,53 +3578,54 @@ PathDiagnosticPieceRef MallocBugVisitor::VisitNode(const ExplodedNode *N,
+        break;
+      }
         case AF_None:
           llvm_unreachable("Unhandled allocation family!");
```
This isn't part of your current patch, but I'd suggest turning this into an assert and return too.

steakhal commented on clang/test/Analysis/MismatchedDeallocator-checker-test.mm:
```diff
@@ -14,6 +14,13 @@
+
+void __attribute((ownership_returns(malloc2))) *my_malloc2(size_t);
+void __attribute((ownership_returns(malloc2))) *my_malloc3(size_t);
+void __attribute((ownership_takes(malloc2, 1))) __attribute((ownership_takes(malloc3, 1))) my_free23(void *);
```
I'm surprised to see that a function can have two different `ownership_takes` attributes.
Are you sure that this is the case? If so, why is that?

Have you considered adding some documentation to the attributes page, and to the Malloc checker to highlight this attribute?

pskrgag replied:
I also find this quite strange, but can't say if there are any real-world examples of function that takes ownership of 2 different classes.

Maybe it would be better to just disallow 2 different classes?

steakhal replied:
To me, it would make more sense to allow at most one of this attribute to be present.

pskrgag replied:
Disallowed 2 ownership_takes with different classes

steakhal commented on clang/test/Analysis/MismatchedDeallocator-checker-test.mm:
```diff
@@ -60,6 +67,41 @@ void testMalloc8() {
+
+void testMalloc12() {
+  int *p = (int *)my_malloc1(sizeof(int));
+  my_free(p); // expected-warning{{Memory allocated by my_malloc1() should be deallocated by function that takes ownership of 'malloc1', not my_free(), which takes ownership of 'malloc'}}
```
It's interesting to see that in a single message we have `my_malloc1()` unquoted; `malloc1` quoted; `my_free()` unquoted; finally `malloc` quoted.
Do you think it's worth it to harmonize these?
Also notice that some of these has parens in the end, some don't.

steakhal replied:
Ah, I see. The phrasing wasn't clear to me. Maybe because we don't have a clear spec for this attribute, thus it's difficult to make out things that have special meaning.
I think we should quote function names, yes.

pskrgag replied:
So I added quotes around all functions and operators. Output looks a way more consistent. Thanks!

steakhal commented on clang/test/Analysis/MismatchedDeallocator-checker-test.mm:
```diff
@@ -14,6 +14,13 @@
 void free(void *);
 void __attribute((ownership_takes(malloc, 1))) my_free(void *);
 
+void __attribute((ownership_returns(malloc1))) *my_malloc1(size_t);
```
What happens if you have a forward declaration to a malloc function without the attribute, then later have another redeclaration with the `ownership_returns` attribute?
A similar case to test would be the opposite: 1st declaration has the attribute, the second redeclaration would not have the attribute.
I suspect that the FunctionDecl of the callees will refer to the last seen declaration of the function, thus, will not recognize the attribute for the second case I mentioned here.
Could you test these cases?

pskrgag replied:
Seems like it works fine. At least, new test case passes. I guess, clang picks the most specific declaration

steakhal replied:
Interesting.

steakhal reviewed (changes requested):
Looks pretty good. Thanks for working on the issue.
I didn't see any major blocking issue with the PR, so good job on that!

Overall, be sure to follow the llvm [coding style](https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly), including using upper case variable names.

Thanks again for working on this :)

pskrgag committed:
clang/sema: disallow more than one 'onweship_takes' with different classes

pskrgag committed:
csa/MallocChecker: quote funtion names in error output

pskrgag committed:
csa/MallocChecker: turn llvm_unreachable into asserts

pskrgag committed:
csa/MallocChecker: add support for custom allocation classes

Add support for custom allocation classes that could be specified by
ownership_returns attribute.

This patch adds basic support, so new class is not contructed anywhere
and handled as an error everywhere.

pskrgag committed:
csa/MismatchedDeallocator: handle custom allocation clasess

Patch introduces support for custom allocation classes in
MismatchedDeallocator.

Patch preserves previous behavior, in which 'malloc' class was handled
as AF_Malloc type.

steakhal commented on clang/test/Sema/attr-ownership.c:
```diff
@@ -24,3 +24,7 @@ void f15(int, int)
 void f16(int *i, int *j) __attribute__((ownership_holds(foo, 1))) __attribute__((ownership_holds(foo, 1))); // OK, same index
 void f17(void*) __attribute__((ownership_takes(__, 1)));
 void f18() __attribute__((ownership_takes(foo, 1)));  // expected-warning {{'ownership_takes' attribute only applies to non-K&R-style functions}}
+
+int f19(void *)
+  __attribute__((ownership_takes(foo, 1)))    // expected-error {{'ownership_takes' attribute class does not match; here it is 'foo'}}
+  __attribute__((ownership_takes(foo1, 1)));  // expected-note {{declared with class 'foo1' here}}
```
What happens if I want to redeclare the function with a different "ownership_takes" attribute in a subsequent line? Will it detect the conflict?
```
int f20(void *) __attribute__((ownership_takes(foo, 1)));
int f20(void *) __attribute__((ownership_takes(foo1, 1))); // redecl with different attr
```

pskrgag replied:
Hmmmmm.... It does not detect. As well as for other attributes. 

I think, we should fix it, but maybe all of them in scope of separate PR?

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -103,14 +103,49 @@ using namespace std::placeholders;
+  void Profile(llvm::FoldingSetNodeID &ID) const {
+    ID.AddInteger(Kind);
+
+    if (Kind == AF_Custom)
+      ID.AddString(CustomName.value());
+  }
```
I think you could also do this unconditionally.

pskrgag replied:
Why? If it's not custom class, then unwrapping `CustomName` would cause an exception, isn't it?

steakhal replied:
Ah indeed. nvm.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -103,14 +103,49 @@ using namespace std::placeholders;
+  explicit AllocationFamily(AllocationFamilyKind kind,
+                            std::optional<StringRef> name = std::nullopt)
+      : Kind(kind), CustomName(name) {
+    assert(kind != AF_Custom || name != std::nullopt);
```
I'd prefer to have a message here to be able to immediately tell what this assert is about when fires.
```suggestion
    assert((kind != AF_Custom || name.has_value()) && "Custom family must specify also the name");
```

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -103,14 +103,49 @@ using namespace std::placeholders;
+  AllocationFamilyKind Kind;
+  std::optional<StringRef> CustomName;
+
+  explicit AllocationFamily(AllocationFamilyKind kind,
+                            std::optional<StringRef> name = std::nullopt)
```
Please use upper camel case for variable names. Shadowing is not a problem for constructor parameters and init-lists, and within the ctor body the semantics are anyways well-defined.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -103,14 +103,49 @@ using namespace std::placeholders;
+      : Kind(kind), CustomName(name) {
+    assert(kind != AF_Custom || name != std::nullopt);
+
+    // Preseve previous behavior when "malloc" class means AF_Malloc
+    if (Kind == AF_Malloc && CustomName) {
+      if (CustomName.value() == "malloc")
+        CustomName = std::nullopt;
+      else
+        Kind = AF_Custom;
+    }
```
Probably I don't understand the semantics of this one.
I was expecting that the `CustomName` field is only used when the `Kind` is `AF_Custom`. Yet, here it appears to be used even for `Kind == AF_Malloc`.

What I originally thought that when `AF_Custom` is specified with `"malloc"`, then it should consider the case as `AF_Malloc` instead.
Could you elaborate here?

pskrgag replied:
Sorry, I over-engineered things. 

I redo a bit. Now callers pass `AF_Custom` with `customName` set, instead of `AF_Malloc`

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1122,7 +1157,7 @@ MallocChecker::performKernelMalloc(const CallEvent &Call, CheckerContext &C,
     SVal ZeroVal = C.getSValBuilder().makeZeroVal(Ctx.CharTy);
     return MallocMemAux(C, Call, Call.getArgExpr(0), ZeroVal, TrueState,
-                        AF_Malloc);
+                        AllocationFamily(AF_Malloc));
```
Now I get what you meant when pushed back on adding `explicit`. I didn't expect this to cause such a widespread effect. Sorry. Now that you made all the necessary changes, we can keep it as-is. You decide if you prefer reverting these back. I'm good either way.

pskrgag replied:
I like current variant more. Just because it makes code more straightforward, I believe

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1490,7 +1532,7 @@ ProgramStateRef MallocChecker::ProcessZeroAllocCheck(
       return State;
     }
   } else
-    llvm_unreachable("not a CallExpr or CXXNewExpr");
+    assert(false && "not a CallExpr or CXXNewExpr");
```
```suggestion
  } else {
    assert(false && "not a CallExpr or CXXNewExpr");
    return nullptr;
  }
```
Just to be on the safe side and we have a NDEBUG build (with assertions disabled), let's avoid undefined behavior by safely backing off.

The same principle should apply to all the other cases when previously we had an `llvm_unreachable`.
This is what I wanted to imply by `assert and return`.

steakhal reviewed (commented):
It looks great! I only had one more thing to test.
Besides that it would be nice to advertise this in the `clang/docs/ReleaseNotes.rst`.
Usually, from that release notes we also link to the checker or the relevant doc pages - in this case it would be nice to link to the docs of this attribute - which does not really elaborate on the attribute and the semantics.

Do you want to extend the scope of this PR to add some minimal docs to the attribute?
If not, that's also fine, we will create a separate ticket for adding them later.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1877,16 +1923,43 @@ static bool didPreviousFreeFail(ProgramStateRef State,
 
+static void printOwnershipTakesList(raw_ostream &os, CheckerContext &C,
+                                    const Expr *E) {
+  if (const CallExpr *CE = dyn_cast<CallExpr>(E)) {
```
When I advocated for early-returns, I wanted something like this:
```c++
const auto *CE = dyn_cast<CallExpr>(E);
if (!CE)
  return;
const FunctionDecl *FD = CE->getDirectCallee();
if (!FD)
  return;
  ...
```
On the same token, the `FD->hasAttrs()` appears to be not strictly necessary as in that case the loop simply wouldn't take any iterations at all, right?

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1877,16 +1923,43 @@ static bool didPreviousFreeFail(ProgramStateRef State,
+
+    // Only one ownership_takes attribute is allowed
+    for (const auto *I : FD->specific_attrs<OwnershipAttr>()) {
+      OwnershipAttr::OwnershipKind OwnKind = I->getOwnKind();
+
+      if (OwnKind != OwnershipAttr::Takes)
```
```suggestion
      if (I->getOwnKind() != OwnershipAttr::Takes)
```
It seems like inlining this variable would make it more readable actually.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1877,16 +1923,43 @@ static bool didPreviousFreeFail(ProgramStateRef State,
+      if (OwnKind != OwnershipAttr::Takes)
+        continue;
+
+      os << ", which takes ownership of " << '\'' << I->getModule()->getName()
+         << '\'';
```
```suggestion
      os << ", which takes ownership of '" << I->getModule()->getName() << '\'';
```
This way it would even fit into a single line.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1877,16 +1923,43 @@ static bool didPreviousFreeFail(ProgramStateRef State,
+    if (!FD->hasAttrs())
+      return;
+
+    // Only one ownership_takes attribute is allowed
```
```suggestion
    // Only one ownership_takes attribute is allowed.
```

steakhal reviewed (approved):
I think it looks great.
Only a few more stylistic remarks are left, but it should be already in a good shape functional-wise. I'll let you decide if you wanna apply my suggestions or not.

One more remark - this time w.r.t. the workflow. Usually, GitHub PRs prefer "merges" over "force-pushes". Whenever you force-push, all the inline remarks could get lost, as it fails to track the lines where the comments were made. Thus, when working with PRs, one usually just "merges" the tracked branch, and just builds more and more commits on top of your PR branch. Once the review is done, one is free to destroy the history and squash the commits, or do some reordering - because there are no more pending comments or remarks left.

pskrgag committed:
csa/MallocChecker: fix style and unexpected shadowing

pskrgag commented:
Thank you for review!

I've applied your suggestions and fixed test failures. Messed up a bit with shadowing in constructor. 

> Usually, GitHub PRs prefer "merges" over "force-pushes". Whenever you force-push, all the inline remarks could get lost, as it fails to track the lines where the comments were made

Oh, ok. I've never really worked with gh workflow, so didn't know that. Thank you for guidance!

steakhal commented:
FYI This PR is still not mentioned in the `clang/docs/ReleaseNotes.rst`.
We should have two entries there:
 1) In [Attribute Changes in Clang](https://github.com/llvm/llvm-project/blob/main/clang/docs/ReleaseNotes.rst#id34), mentioning that calls can have at most one `ownership_return` attribute.
 2) In [Static Analyzer > New features](https://github.com/llvm/llvm-project/blob/main/clang/docs/ReleaseNotes.rst#id47), that the MallocChecker now checks this attribute, and you can read the docs for the checker, and the attribute at the links.

pskrgag commented:
> and you can read the docs for the checker, and the attribute at the links.

I can't find any docs for this attribute. As I mentioned, I will fill new issues to fix couple of frontend issues and after that we can write down correct semantics of these attrs.

Do you mean that link https://clang.llvm.org/docs/AttributeReference.html#ownership-holds-ownership-returns-ownership-takes?

pskrgag committed:
add release note

steakhal commented:
> > and you can read the docs for the checker, and the attribute at the links.
> 
> I can't find any docs for this attribute. As I mentioned, I will fill new issues to fix couple of frontend issues and after that we can write down correct semantics of these attrs.

Thats fine.

> Do you mean that link https://clang.llvm.org/docs/AttributeReference.html#ownership-holds-ownership-returns-ownership-takes?

Yes. Its generated from a tablegen file. Thats why you probably didnt find it.
Look for files with the "td" extension, and  "attr" in name. May have a different casing though. Watch out.

pskrgag committed:
add example for mismatched deallocation class in documentation

pskrgag commented:
Rebased on top of d89f3e8df3160b3afc07bc742c81aa4738ea9646

steakhal commented:
Should I merge this PR, or you want to wait some?

pskrgag commented:
I think, it's good time to merge, since v19 release branch was created.

If you want to wait for others to review, it's also fine.
